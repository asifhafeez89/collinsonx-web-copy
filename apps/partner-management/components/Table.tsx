import {
  TriangleDown,
  TriangleUp,
} from '@collinsonx/design-system/components/table';
import { Box, Table as MantineTable } from '@collinsonx/design-system/core';
import styled from '@collinsonx/design-system/styled';
import { Table as ReactTable, flexRender } from '@tanstack/react-table';
import { PageType, bookingPageConfig } from 'config/booking';

const CustomTable = ({
  headerBg,
  ...props
}: {
  headerBg: string;
  children: JSX.Element[];
}) => <MantineTable {...props} />;

const StyledTable = styled(CustomTable)`
  color: #333333;
  border: 1px solid #e9e9e9;
  border-top: 0;
  tr {
    height: 78px;
  }
  thead tr {
    height: 50px;
    background: ${(props) => props.headerBg};
  }
  tbody {
    tr:nth-of-type(even) {
      background: #f9f9f9;
    }
    tr:hover {
      background: #ededed;
    }
  }

  thead th + th,
  thead tr th:last-child {
    border-left: 1px solid #dad1bb;
  }

  tbody tr td {
    border-top: 0;
  }
`;

interface BookingsTableProps<T> {
  table: ReactTable<T>;
  type: PageType;
  widthColMap?: Record<string, number>;
}

function Table<T extends unknown>({
  table,
  type,
  widthColMap = {},
}: BookingsTableProps<T>) {
  return (
    <StyledTable
      headerBg={bookingPageConfig[type as keyof typeof bookingPageConfig].color}
    >
      <thead>
        {table.getHeaderGroups().map((headerGroup) => (
          <tr key={headerGroup.id}>
            {headerGroup.headers.map((header) => (
              <th
                key={header.id}
                style={{
                  width:
                    widthColMap[header.id as keyof typeof widthColMap] ??
                    'auto',
                }}
              >
                <Box
                  {...{
                    sx: {
                      cursor: header.column.getCanSort() ? 'pointer' : 'auto',
                    },
                    onClick: header.column.getToggleSortingHandler(),
                  }}
                >
                  {flexRender(
                    header.column.columnDef.header,
                    header.getContext()
                  )}

                  <span style={{ paddingLeft: 8 }}>
                    {{
                      asc: <TriangleUp color="#000" />,
                      desc: <TriangleDown color="#000" />,
                    }[header.column.getIsSorted() as string] ?? null}
                  </span>
                </Box>
              </th>
            ))}
          </tr>
        ))}
      </thead>
      <tbody>
        {table.getRowModel().rows.map((row) => (
          <tr key={row.id}>
            {row.getVisibleCells().map((cell) => (
              <td key={cell.id}>
                {flexRender(cell.column.columnDef.cell, cell.getContext())}
              </td>
            ))}
          </tr>
        ))}
      </tbody>
    </StyledTable>
  );
}

export default Table;
