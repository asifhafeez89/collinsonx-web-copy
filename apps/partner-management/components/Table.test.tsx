import Table from './Table';
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import {
  Table as ReactTable,
  RowModel,
  useReactTable,
} from '@tanstack/react-table';

jest.mock('@collinsonx/design-system/components/table/index', () => {
  return {
    __esModule: true,
    TriangleDown: () => [],
    TriangleUp: () => [],
  };
});
jest.mock('@tanstack/react-table', () => {
  return {
    __esModule: true,
    useReactTable: () => ({
      getHeaderGroups: () => {
        return [];
      },
      getRowModel: () => {
        return {
          rows: [
            {
              id: '123',
              getVisibleCells: () => {
                return [
                  {
                    id: '23',
                    column: {
                      columnDef: {
                        cell: 2,
                      },
                    },
                    getContext: () => [],
                  },
                ];
              },
            },
          ],
        };
      },
    }),
    flexRender: () => [],
  };
});
describe('<type />', () => {
  it('should render', () => {
    const table = useReactTable({
      data: [
        {
          __typename: 'Booking',
          bookedFrom: '2023-08-23 10:44:24.431',
          bookedTo: '2023-08-23 12:44:24.431',
          createdAt: '2023-08-21T09:44:24.624Z',
          type: 'RESERVATION',
          metadata: {},
          id: '64e331f825b764f3c583d509',
          reference: 'K3N6H6',
          guestAdultCount: 1,
          guestChildrenCount: 1,
          guestInfantCount: 1,
          status: 'DECLINED',
          updatedAt: '2023-09-05T13:24:22.772Z',
          consumer: {
            __typename: 'Consumer',
            emailAddress:
              'automationconsumertest@collinsonxteam.testinator.com',
            firstName: 'Automation',
            fullName: 'Automation Consumer',
            id: 'd1a9234f-3210-4673-8d15-0d2f360b639e',
          },
          experience: {
            __typename: 'Experience',
            id: '1ccc3807-a7ed-5a3a-ada8-fd37ac1ab941',
            loungeName: 'Clubrooms Birmingham',
          },
          _id: 'K3N6H6',
          arrivalDate: '2023-08-23',
          arrivalTime: '10:44',
        },
      ],
      getCoreRowModel: function (table: ReactTable<any>): () => RowModel<any> {
        throw new Error('Function not implemented.');
      },
      columns: [
        {
          id: 'fullName',
          header: 'Customer name',
          accessorKey: 'consumer.fullName',
        },
        {
          id: '_id',
          header: 'Booking reference',
          accessorKey: '_id',
        },
        {
          header: 'Type',
          id: 'type',
          accessorKey: 'type',
        },
        {
          header: 'Arrival date',
          id: 'arrivalDate',
          accessorKey: 'arrivalDate',
        },
        {
          header: 'Arrival time',
          id: 'arrivalTime',
          accessorKey: 'arrivalTime',
        },
        {
          header: 'Guests',
          id: 'guestCount',
          accessorKey: 'guestCount',
        },
      ],
    });
    const component = render(<Table table={table} type={'pending'} />);

    expect(component).toMatchSnapshot();
  });
});
