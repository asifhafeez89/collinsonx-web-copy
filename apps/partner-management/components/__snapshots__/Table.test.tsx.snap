// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<type /> should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": .emotion-1 {
  color: #333333;
  border: 1px solid #e9e9e9;
  border-top: 0;
}

.emotion-1 tr {
  height: 78px;
}

.emotion-1 thead tr {
  height: 50px;
  background: #FFF3BF;
}

.emotion-1 tbody tr:nth-of-type(even) {
  background: #f9f9f9;
}

.emotion-1 tbody tr:hover {
  background: #ededed;
}

.emotion-1 thead th+th,
.emotion-1 thead tr th:last-child {
  border-left: 1px solid #dad1bb;
}

.emotion-1 tbody tr td {
  border-top: 0;
}

.emotion-2 {
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  width: 100%;
  border-collapse: collapse;
  caption-side: top;
  color: #000;
  line-height: 1.55;
}

.emotion-2>caption {
  margin-top: 0;
  margin-bottom: 0.625rem;
  font-size: 0.875rem;
  color: #868e96;
}

.emotion-2>thead>tr>th,
.emotion-2>tfoot>tr>th,
.emotion-2>tbody>tr>th {
  text-align: left;
  font-weight: bold;
  color: #495057;
  font-size: 0.875rem;
  padding: 0.4375rem 0.625rem;
}

.emotion-2>thead>tr>th {
  border-bottom: 0.0625rem solid #dee2e6;
}

.emotion-2>tfoot>tr>th,
.emotion-2>tbody>tr>th {
  border-top: 0.0625rem solid #dee2e6;
}

.emotion-2>tbody>tr>td {
  padding: 0.4375rem 0.625rem;
  border-top: 0.0625rem solid #dee2e6;
  font-size: 0.875rem;
}

.emotion-2>tbody>tr:first-of-type>td,
.emotion-2>tbody>tr:first-of-type>th {
  border-top: none;
}

.emotion-2>thead>tr>th,
.emotion-2>tbody>tr>td {
  border-right: none;
}

.emotion-2>thead>tr>th:last-of-type,
.emotion-2>tbody>tr>td:last-of-type {
  border-right: none;
  border-left: none;
}

.emotion-2>tbody>tr>th {
  border-right: none;
}

.emotion-2[data-striped]>tbody>tr:nth-of-type(odd) {
  background-color: #f8f9fa;
}

@media (hover: hover) {
  .emotion-2[data-hover]>tbody>tr:hover {
    background-color: #f1f3f5;
  }
}

@media (hover: none) {
  .emotion-2[data-hover]>tbody>tr:active {
    background-color: #f1f3f5;
  }
}

<body>
    <div>
      <table
        class="emotion-0 emotion-1 emotion-2"
      >
        <thead />
        <tbody>
          <tr>
            <td />
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": .emotion-1 {
  color: #333333;
  border: 1px solid #e9e9e9;
  border-top: 0;
}

.emotion-1 tr {
  height: 78px;
}

.emotion-1 thead tr {
  height: 50px;
  background: #FFF3BF;
}

.emotion-1 tbody tr:nth-of-type(even) {
  background: #f9f9f9;
}

.emotion-1 tbody tr:hover {
  background: #ededed;
}

.emotion-1 thead th+th,
.emotion-1 thead tr th:last-child {
  border-left: 1px solid #dad1bb;
}

.emotion-1 tbody tr td {
  border-top: 0;
}

.emotion-2 {
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  width: 100%;
  border-collapse: collapse;
  caption-side: top;
  color: #000;
  line-height: 1.55;
}

.emotion-2>caption {
  margin-top: 0;
  margin-bottom: 0.625rem;
  font-size: 0.875rem;
  color: #868e96;
}

.emotion-2>thead>tr>th,
.emotion-2>tfoot>tr>th,
.emotion-2>tbody>tr>th {
  text-align: left;
  font-weight: bold;
  color: #495057;
  font-size: 0.875rem;
  padding: 0.4375rem 0.625rem;
}

.emotion-2>thead>tr>th {
  border-bottom: 0.0625rem solid #dee2e6;
}

.emotion-2>tfoot>tr>th,
.emotion-2>tbody>tr>th {
  border-top: 0.0625rem solid #dee2e6;
}

.emotion-2>tbody>tr>td {
  padding: 0.4375rem 0.625rem;
  border-top: 0.0625rem solid #dee2e6;
  font-size: 0.875rem;
}

.emotion-2>tbody>tr:first-of-type>td,
.emotion-2>tbody>tr:first-of-type>th {
  border-top: none;
}

.emotion-2>thead>tr>th,
.emotion-2>tbody>tr>td {
  border-right: none;
}

.emotion-2>thead>tr>th:last-of-type,
.emotion-2>tbody>tr>td:last-of-type {
  border-right: none;
  border-left: none;
}

.emotion-2>tbody>tr>th {
  border-right: none;
}

.emotion-2[data-striped]>tbody>tr:nth-of-type(odd) {
  background-color: #f8f9fa;
}

@media (hover: hover) {
  .emotion-2[data-hover]>tbody>tr:hover {
    background-color: #f1f3f5;
  }
}

@media (hover: none) {
  .emotion-2[data-hover]>tbody>tr:active {
    background-color: #f1f3f5;
  }
}

<div>
    <table
      class="emotion-0 emotion-1 emotion-2"
    >
      <thead />
      <tbody>
        <tr>
          <td />
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
